### DISCUSSION TITLE
Mapper using another mapper

### DISCUSSION THREAD

**Discussion by jakoss** [mcarleio/konvert#11] [Ideas]
In MapStruct we can generate a mapper, let's call it `TestMapper` and then we can use mappings generated by this mapper in another mapper by `@Mapper(using = [TestMapper::class])`. Is there a way to do the same in Konvert? If not - i think this is a major use case that should be supported

**Comment by mcarleio** [comment]
If they are being processed in the same KSP run, then this is already working without any extra stuff. But if the `TestMapper` was compiled before, then at the moment this is not possible. I already thought about that, but nothing implemented yet.

**Comment by mcarleio** [comment]
With the latest release 2.0.0 it now is possible to use generated functions from other modules or libraries. There is no need to define which generated functions to use, as Konvert now additionally generates some files in META-INF to document generated functions and during processing collects all those files and register the functions behind as normal `TypeConverter`.

Please try it out and give feedback, if it is working as you would expect.

**Reply by jakoss** [reply]
I tried that and i have some issue. In module 'core' i have this:
```kotlin
@Konverter
interface DateTimeConverters {
    fun asOffsetDateTime(source: LocalDateTime): OffsetDateTime = source.atOffset(localZoneOffset)

    companion object {
        val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
    }
}
```

Then in feature module i have this:
```kotlin
@Konverter
interface WeatherMapper : DateTimeConverters {
    fun fromDto(weatherDto: CurrentWeatherDto): CurrentWeather
    fun toDto(weather: CurrentWeather): CurrentWeatherDto
}
```

Generated code for this looks like this:
```kotlin
@ContributesBinding(scope = ApplicationScope::class)
public class WeatherMapperImpl : WeatherMapper {
  @Inject
  public constructor()

  @GeneratedKonverter(priority = 5_000)
  override fun fromDto(weatherDto: CurrentWeatherDto): CurrentWeather = CurrentWeather(
    temperature = weatherDto.temperature,
    windspeed = weatherDto.windspeed,
    time = weatherDto.time.toLocalDateTime()
  )

  @GeneratedKonverter(priority = 5_000)
  override fun toDto(weather: CurrentWeather): CurrentWeatherDto = CurrentWeatherDto(
    temperature = weather.temperature,
    windspeed = weather.windspeed,
    time = this.asOffsetDateTime(source = weather.time)
  )
}
``` 

So this is great, looks like magic :) BUT, in top level app module i have this error:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeDebugJavaResource'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.MergeJavaResWorkAction
   > 2 files found with path 'META-INF/konvert/io.mcarle.konvert.api.Konvert' from inputs:
      - C:\Projects\ArchitectureTemplate\data\weather\impl\build\intermediates\library_java_res\debug\res.jar
      - C:\Projects\ArchitectureTemplate\core\infrastructure\build\intermediates\library_java_res\debug\res.jar
     Adding a packagingOptions block may help, please refer to
     https://developer.android.com/reference/tools/gradle-api/8.0/com/android/build/api/dsl/ResourcesPackagingOptions
     for more information
```

So this seems to be an issue with multiple modules having konvert processor applied. 

As usual, you can test this out on my repo: https://github.com/jakoss/AndroidArchitectureTemplate (just bump konvert version in ./gradle/libs.versions.toml and try to run the app)

**Reply by mcarleio** [reply]
Could you try (in your app module) to exclude META-INF/konvert/* from being packaged into the apk? I have not installed Android Studio and therefore could not test myself, but am confident, that this will solve the issue you have.

It could be looking like this:

```
android {
   packagingOptions {
       resources.excludes.add("META-INF/konvert/*")
   }
}
```

Alternative to `exclude` is to `merge` them per file, which would probably look like this:

```
android {
   packagingOptions {
       resources.merges.add("META-INF/konvert/io.mcarle.konvert.api.Konvert")
       resources.merges.add("META-INF/konvert/io.mcarle.konvert.api.KonvertTo")
       resources.merges.add("META-INF/konvert/io.mcarle.konvert.api.KonvertFrom")
   }
}
```

But as this file is not used during runtime (only during KSP processor "runtime" during build) you can probably simply exclude them.

**Reply by jakoss** [reply]
Adding an exclude workes like a charm, nice! I have to see what are the magic of META-INF you did there :)

**Reply by mcarleio** [reply]
In a nutshell, during generation of kotlin code, one or more of these three files will be generated. They include FQN to the generated functions. During KSP processing, all of those files on the classpath are loaded and each of the functions in there are registered as `TypeConverter` and can therefore be used by `Konvert` during processing.

In your case, I think you actually do not use this feature, as there is no code being generated for the interface `DateTimeConverters`. I think you can even remove the `@Konverter` annotation there and everything should keep working.

**Reply by jakoss** [reply]
Oh yea, it wasn't possible before. Now my custom converter in interface can be "Konvert agnostic" and i don't have to run the processor on my core module. That's nice :)
